[HWKINVENT-88] Fix stable ID computation in REST

metlos opened the pull request.

metlos committed:
[HWKINVENT-88] Replace the stable ID with just canonical path.

The stable ID computation was failing for child resources.

This required replacing the "URL-like" source and target paths
in relationship JSON with the canonical path, too.

metlos committed:
[HWKINVENT-88] Fix CanonicalPath.ids()

For all "terminal" entity types, the id is taken from last segment, no
need to go down from the path's root.

For resource, the fact that resources are hierarchical is taken into
account returning a relative path of the resource hierarchy rather than
just ID.

jkremser commented on api/src/main/java/org/hawkular/inventory/api/model/CanonicalPath.java:
```diff
@@ -278,27 +278,43 @@ public String getEnvironmentId() {
 
         public String getMetricTypeId() {
-            return idIfTypeCorrect(getRoot().down(), MetricType.class);
+            return idIfTypeCorrect(CanonicalPath.this, Resource.class);
```
`s/Resource/Metric(Type)/` ?

metlos replied:
Oh, the joys of copy paste :) Thanks for spotting it.

jkremser commented on api/src/main/java/org/hawkular/inventory/api/model/CanonicalPath.java:
```diff
@@ -278,27 +278,43 @@ public String getEnvironmentId() {
 
         public String getResourceTypeId() {
-            return idIfTypeCorrect(getRoot().down(), ResourceType.class);
+            return idIfTypeCorrect(CanonicalPath.this, Resource.class);
```
This looks suspicious too. `getResourceTypeId` -> `ResourceType.class` (?)

jkremser commented on api/src/main/java/org/hawkular/inventory/api/model/CanonicalPath.java:
```diff
@@ -278,27 +278,43 @@ public String getEnvironmentId() {
-            String id = idIfTypeCorrect(getRoot().down(2), Metric.class);
-
-            return id != null ? id : idIfTypeCorrect(getRoot().down(3), Metric.class);
+            return idIfTypeCorrect(CanonicalPath.this, Resource.class);
```
---||---

metlos committed:
[HWKINVENT-88] Fixing child resource inspection and descend.

These code-paths were not tested at all and fixing them (and adding
tests) uncovered a number of subtle bugs in path handling and query
creation.

metlos committed:
[HWKINVENT-88] Oops, silly copy&paste error.

metlos committed:
[HWKINVENT-88] Fix compilation errors.
